lldb/include/lldb/Target/Target.h

```diff
@@ -1692,6 +1692,20 @@ class Target : public std::enable_shared_from_this<Target>,
   }
 };
 
+/// The private implementation backing SBLock.
+struct APILock {
+  APILock(std::recursive_mutex &mutex) : lock(mutex) {}
+  std::lock_guard<std::recursive_mutex> lock;
+};
+
+/// The private implementation used by SBLock to hand out the target API mutex.
+/// It has a TargetSP to ensure the lock cannot outlive the target.
+struct TargetAPILock : public APILock {
```

Comment:
why do we have a base class? Can't we simply merge `TargetAPILock` into `APILock`?

Reply:
I consciously did it this way to avoid giving the impression that `SBLock` can only be used for the target API lock. The latter is the only user, but if in the future there's another lock we want to expose through the SB API, it can inherit from this class.

Reply:
So, if all you want to do with this is to hold a shared pointer, then this can be done in a simpler way. You can have the class take a `std::shared_ptr<mutex>` and call this using the shared_ptr [aliasing constructor](https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr) from the shared pointer of your choice (e.g. `TargetAPILock(std::shared_ptr<mutex>(target_sp, target_sp->m_mutex))`)

Reply:
This doesn't quite work here because it's a C++20 extension and we don't actually need a pointer to the `mutex`, we need the lock. If we did build with C++20 support I could change `TargetAPILock` to hold onto a shared_ptr for the mutex, to make it a little bit more obvious, but in the absence of that I think the current approach is the best we can do.

Reply:
I don't think that's true. The overload (number 8 in the cppreference page) which takes a rvalue reference is a c++20 thing, but the version which takes a `const&` has been around since forever. I'm pretty sure of that because we're [already using](https://github.com/llvm/llvm-project/blob/1f844952558b041ff1b4c27eed7b81c15841ee84/lldb/include/lldb/Utility/SharedCluster.h#L48) this functionality.

Reply:
I was indeed looking at (8). But even if it's available it still takes an existing reference/pointer, while what I need is to create a new object the (lock_guard) based on the pointer. Unless you mean changing the shared_ptr in the TargetAPILock to hold onto the mutex that way, which is indeed possible (but doesn't eliminate the class, as I had hoped).

Reply:
What I meant was that you don't need the *Target*APILock class as this can be implemented generically in the APILock base class (although it looks like my snippet https://github.com/llvm/llvm-project/pull/131404/files#r1998175315 still calls this TargetAPILock). I.e., you could do something like:

```
class APILock {
  APILock(std::shared_ptr<std::recursive_mutex> mutex) : m_mutex(std::move(mutex)), m_lock(*m_mutex) {}

private:
  std::shared_ptr<std::recursive_mutex> m_mutex;
  std::lock_guard<std::recursive_mutex> m_lock;
};
```

and this can be used with anything that can turn itself into a shared pointer (using the aliasing constructor)

Reply:
Ah, gotcha, yeah that makes sense. Thanks for walking me through it.